### PR TITLE
Use fsspec to open file in eager read_* functions

### PR DESCRIPTION
**Pull request by olivier-lacroix**
Hi @ritchie46 ,

This PR
- adds a dependency on [fsspec](https://github.com/intake/filesystem_spec);
- relies on fsspec to open files passed as string arguments to polars read_* functions;
- defaults to inferring file compression format

This offloads to fsspec the handling of (de)compression, and remote protocls (e.g. s3://, gs://)

### CODE REVIEW

**Comment by olivier-lacroix** [comment]
@ritchie46 , while doing this, I had a look at `frame.py`. It appears all read_* methods of DataFrame but read_csv allow to switch between pyarrow and polars paths. This seems a bit redundant as this logic is already present in the `functions.py` module. Shouldn't pyarrow paths be removed from the DataFrame methods? or alternatively the switch between both paths removed from `functions.py` ?

**Comment by ritchie46** [comment]
@olivier-lacroix , thanks for the PR. I still have some questions. 

I don't want to add a new dependency, so can this be made optional. Is it correct that ffspec now opens all files, even paths? This is not desired, as the rust code is a lot slower if it has to deal with python objects. There are also fast paths if we know the full path as string.

**Comment by olivier-lacroix** [comment]
@ritchie46 ,

- I can make it optional yes.
- Indeed. I saw absolutely no performance difference when passing a path or a python file handle to polars.read_csv , but maybe I am missing something ?

```python
import polars as pl
import fsspec

# filename is a CSV file

pl.read_csv(filename)
# 1.62 s ± 21.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

with fsspec.open(filename) as f:
    pl.read_csv(f)
# 1.62 s ± 17.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

```

**Comment by ghuls** [comment]
I think in the last comparison you are comparing fsspec twice, as pyarrow will also use it when installed. Testing with `use_pyarrow=False` will actually test if opening the file directly in rust or fsspec has a big impact.

**Comment by olivier-lacroix** [comment]
@ghuls , indeed. and it does:

```python

pl.read_csv(filename,use_pyarrow=False)
# 1.55 s ± 17.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

with fsspec.open(filename) as f:
    pl.read_csv(f,use_pyarrow=False)
# 5.02 s ± 13.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
@ritchie46 , I will update the PR to return a string for a local, non compressed file

**Comment by olivier-lacroix** [comment]
@ritchie46, PR updated accordingly.

**Comment by ritchie46** [comment]
> @ritchie46, PR updated accordingly.
> 
> Note that non local files now (e.g. over http(s)) require fsspec to work properly.

I think that's a regression. The python standard library supports doing http requests. I think we should fallback to previous logic if `ffspec` is not installed.

**Comment by ghuls** [comment]
The performance impact from ffspec for `read_csv(..., use_pyarrow=False)` is due the loss of multithreaded processing.

```
In [48]: import pyarrow as pa

In [49]: pa.set_cpu_count(1)

# Read directly with pyarrow (1 thread).
In [50]: %%time
    ...: df = pl.read_csv(filename, use_pyarrow=True)
    ...: 
    ...: 
CPU times: user 10.7 s, sys: 10.1 s, total: 20.8 s
Wall time: 18.6 s

In [51]: pa.set_cpu_count(16)

# Read directly with pyarrow (16 threads).
In [52]: %%time
    ...: df = pl.read_csv(filename, use_pyarrow=True)
    ...: 
    ...: 
CPU times: user 11.9 s, sys: 11.4 s, total: 23.3 s
Wall time: 5.61 s

In [53]: pa.set_cpu_count(1)

# Read directly with polars (16 threads).
In [54]: %%time
    ...: df = pl.read_csv(filename, use_pyarrow=False, n_threads=16)
    ...: 
    ...: 
avg line length: 41.33008
std. dev. line length: 4.8249497
initial row estimate: 111854864
no. of chunks: 13654 processed by: 16 threads at 1 chunk/thread
CPU times: user 30.3 s, sys: 21.6 s, total: 51.9 s
Wall time: 8.51 s


# Read directly with polars (1 thread).
In [55]: %%time
    ...: df = pl.read_csv(filename, use_pyarrow=False, n_threads=1)
    ...: 
    ...: 
avg line length: 41.33008
std. dev. line length: 4.8249497
initial row estimate: 111854864
no. of chunks: 13654 processed by: 1 threads at 1 chunk/thread
CPU times: user 25.9 s, sys: 10.6 s, total: 36.5 s
Wall time: 36.3 s


# Read from python open object with polars (1 thread).
In [56]: %%time
    ...: with open(filename, mode='rb') as f:
    ...:     pl.read_csv(f, use_pyarrow=False, n_threads=1)
    ...: 
avg line length: 41.33008
std. dev. line length: 4.8249497
initial row estimate: 111854864
no. of chunks: 13654 processed by: 1 threads at 1 chunk/thread
CPU times: user 30 s, sys: 21 s, total: 51 s
Wall time: 51.1 s

# Read from python open object with with polars (16 threads asked, but will only use 1).
In [57]: %%time
    ...: with open(filename, mode='rb') as f:
    ...:     pl.read_csv(f, use_pyarrow=False, n_threads=16)
    ...: 
avg line length: 41.33008
std. dev. line length: 4.8249497
initial row estimate: 111854864
no. of chunks: 13654 processed by: 1 threads at 1 chunk/thread
CPU times: user 30.7 s, sys: 21.4 s, total: 52.1 s
Wall time: 52.2 s

In [59]: import fsspec

# Read from fsspec open object with with polars (16 threads asked, but will only use 1).
In [60]: %%time
    ...: with fsspec.open(filename, mode='rb') as f:
    ...:     pl.read_csv(f, use_pyarrow=False, n_threads=16)
    ...: 
avg line length: 41.33008
std. dev. line length: 4.8249497
initial row estimate: 111854864
no. of chunks: 13654 processed by: 1 threads at 1 chunk/thread
CPU times: user 29.8 s, sys: 20.6 s, total: 50.5 s
Wall time: 50.6 s
```

**Comment by olivier-lacroix** [comment]
@ritchie46 , fair enough. Will do.

**Comment by olivier-lacroix** [comment]
@ghuls that makes sense thanks! The performance impact of opening files via python is really high. Relying on python to access non-local files is better than nothing, but being able to do so natively in rust seems like the real LT solution.

**Comment by ritchie46** [comment]
@olivier-lacroix There are some conflicts with master.

**Comment by olivier-lacroix** [comment]
Yup. Will rebase over the WE

**Comment by olivier-lacroix** [comment]
@ritchie46 , I have rebased on master.
- fsspec is optional, and will be used for non local or compressed files if present.
- when it is not, reading from http(s) still works
Let me know what you think.

**Comment by ritchie46** [comment]
>  fsspec is optional, and will be used for non local or compressed files if present.
    when it is not, reading from http(s) still works

Nice, I think that's good! There still seem to be some inconsistencies in the typing. Mypy is failing in CI.

**Comment by olivier-lacroix** [comment]
@ritchie46, yes, I am on it. Type hints and optional import is not trivial. But I will get it done!

**Comment by olivier-lacroix** [comment]
I think typing is OK now. mypy still returns some errors but unrelated to this PR I think ?

```
polars/ffi.py:39: error: Module has no attribute "int_asbuffer"
polars/ffi.py:39: error: Unsupported operand types for * ("int" and "Callable[[_ArrayOrScalarCommon], int]")
polars/series.py:156: error: Incompatible types in assignment (expression has type "dtype[Any]", variable has type "Optional[Type[DataType]]")
polars/series.py:222: error: Argument 1 to "_find_first_non_none" has incompatible type "Optional[List[Any]]"; expected "List[Optional[Any]]"
polars/series.py:304: error: Argument 2 to "Series" has incompatible type "Sequence[Any]"; expected "Union[Series, ndarray, List[Any], None]"
polars/series.py:314: error: Argument 2 to "Series" has incompatible type "Sequence[Any]"; expected "Union[Series, ndarray, List[Any], None]"
polars/series.py:324: error: Argument 2 to "Series" has incompatible type "Sequence[Any]"; expected "Union[Series, ndarray, List[Any], None]"
polars/series.py:334: error: Argument 2 to "Series" has incompatible type "Sequence[Any]"; expected "Union[Series, ndarray, List[Any], None]"
polars/series.py:344: error: Argument 2 to "Series" has incompatible type "Sequence[Any]"; expected "Union[Series, ndarray, List[Any], None]"
polars/series.py:354: error: Argument 2 to "Series" has incompatible type "Sequence[Any]"; expected "Union[Series, ndarray, List[Any], None]"
polars/series.py:392: error: Argument 1 to "__call__" of "ufunc" has incompatible type "Series"; expected "Union[Union[int, float, complex, str, bytes, generic], Sequence[Union[int, float, complex, str, bytes, generic]], Sequence[Sequence[Any]], _SupportsArray]"
polars/series.py:444: error: Argument 2 to "__call__" of "ufunc" has incompatible type "Series"; expected "Union[Union[int, float, complex, str, bytes, generic], Sequence[Union[int, float, complex, str, bytes, generic]], Sequence[Sequence[Any]], _SupportsArray]"
polars/series.py:1122: error: Argument 1 to "append" of "list" has incompatible type "ndarray"; expected "Number"
polars/series.py:1905: error: Incompatible return value type (got "Type[Float64]", expected "Union[DataType, ndarray]")
polars/series.py:1907: error: Incompatible return value type (got "Type[Int64]", expected "Union[DataType, ndarray]")
polars/series.py:1909: error: Incompatible return value type (got "Type[Utf8]", expected "Union[DataType, ndarray]")
polars/series.py:1911: error: Incompatible return value type (got "Type[Boolean]", expected "Union[DataType, ndarray]")
polars/series.py:1913: error: Incompatible return value type (got "Type[polars.datatypes.List]", expected "Union[DataType, ndarray]")
polars/series.py:1915: error: Incompatible return value type (got "Type[ndarray]", expected "Union[DataType, ndarray]")
polars/frame.py:88: error: Argument 2 to "Series" has incompatible type "Sequence[Any]"; expected "Union[Series, ndarray, List[Any], None]"
polars/frame.py:730: error: Argument 2 to "Series" has incompatible type "Sequence[Any]"; expected "Union[Series, ndarray, List[Any], None]"
polars/frame.py:2153: error: "polars.lazy.Expr*" object is not iterable
polars/lazy/__init__.py:1446: error: Argument 2 to "Series" has incompatible type "Expr"; expected "Union[Series, ndarray, List[Any], None]"
Found 23 errors in 4 files (checked 9 source files)
```

File: py-polars/polars/functions.py [hunk 1]
```diff
@@ -1,85 +1,125 @@
...
+            if local and not compressed:
+                return managed_file(make_path_posix(file))
+            compression = kwargs.pop("compression", "infer")
+            return fsspec.open(file, compression=compression, **kwargs)
+        if file.startswith("http"):
+            return _process_http_file(file)
+        return managed_file(file)
+    if isinstance(file, list) and bool(file) and all(isinstance(f, str) for f in file):
+        if WITH_FSSPEC:
+            compressed = any(infer_compression(f) is not None for f in file)
```
Comment (olivier-lacroix) [inline py-polars/polars/functions.py:113]: @ritchie46 Compression and protocol is checked for all files in the list. we could maybe check for the first one only? I can't really think of any case where these would be mixed.

**Comment by ritchie46** [comment]
> I think typing is OK now. mypy still returns some errors but unrelated to this PR I think ?

I don't think so, as mypy is succesul in master. :)

**Comment by olivier-lacroix** [comment]
@ritchie46 , this is strange. running mypy (version 0.910) in the py-polars directory yields these 23 errors in master for me.

**Comment by olivier-lacroix** [comment]
@ritchie46 , it seems mypy outcomes depends on versions. I do not have the error reported in CI locally. I'll try and match versions to fix this.

**Comment by olivier-lacroix** [comment]
@ritchie46 , thx for running CI. I missed one type hint issue that I just fixed. hopefully, this should be OK now.

**Comment by ritchie46** [comment]
Thanks @olivier-lacroix. LGTM. :+1:

**Pull request merged**